src/OpenTelemetry.Instrumentation.ConfluentKafka/OpenTelemetryConsumeResultExtensions.cs

```diff
@@ -0,0 +1,184 @@
+    /// Consumes a message and creates <see href="https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md#span-kind">a process span</see> embracing the <see cref="ConsumeAndProcessMessageHandler{TKey,TValue}"/>.
+    /// </summary>
+    /// <param name="consumer">The <see cref="IConsumer{TKey,TValue}"/>.</param>
+    /// <param name="handler">A <see cref="ConsumeAndProcessMessageHandler{TKey,TValue}"/>.</param>
+    /// <param name="cancellationToken">An optional <see cref="CancellationToken"/>.</param>
+    /// <typeparam name="TKey">The type of key of the <see cref="ConsumeResult{TKey,TValue}"/>.</typeparam>
+    /// <typeparam name="TValue">The type of value of the <see cref="ConsumeResult{TKey,TValue}"/>.</typeparam>
+    /// <returns>A <see cref="ValueTask"/>.</returns>
+    public static async ValueTask<ConsumeResult<TKey, TValue>?> ConsumeAndProcessMessageAsync<TKey, TValue>(
+        this IConsumer<TKey, TValue> consumer,
+        ConsumeAndProcessMessageHandler<TKey, TValue> handler,
+        CancellationToken cancellationToken)
+    {
+#if NETFRAMEWORK
+        if (consumer == null)
+        {
+            throw new ArgumentNullException(nameof(consumer));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(consumer);
+#endif
+
+        if (consumer is not InstrumentedConsumer<TKey, TValue> instrumentedConsumer)
+        {
+            throw new ArgumentException("Invalid consumer type.", nameof(consumer));
+        }
+
+#if NETFRAMEWORK
+        if (handler is null)
+        {
+            throw new ArgumentNullException(nameof(handler));
+        }
+#else
+        ArgumentNullException.ThrowIfNull(handler);
+#endif
+
+        var consumeResult = instrumentedConsumer.Consume(cancellationToken);
+
+        if (consumeResult?.Message == null || consumeResult.IsPartitionEOF)
+        {
+            return consumeResult;
+        }
+
+        Activity? processActivity = null;
+        if (TryExtractPropagationContext(consumeResult, out var propagationContext))
+        {
+            processActivity = StartProcessActivity(
+                propagationContext,
+                consumeResult.TopicPartitionOffset,
+                consumeResult.Message.Key,
+                instrumentedConsumer.Name,
+                instrumentedConsumer.GroupId!);
+        }
+        else
+        {
+            processActivity = StartProcessActivity(
+                default,
+                consumeResult.TopicPartitionOffset,
+                consumeResult.Message.Key,
+                instrumentedConsumer.Name,
+                instrumentedConsumer.GroupId!);
+        }
+
+        try
+        {
+            await handler(consumeResult, processActivity, cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+            processActivity?.SetStatus(Status.Error);
+            processActivity?.SetTag(SemanticConventions.AttributeErrorType, ex.GetType().Name);
```

Comment:
nit:

```suggestion
            processActivity?.SetTag(SemanticConventions.AttributeErrorType, ex.GetType().FullName);
```

that's what .NET does and we're yet to properly document it - https://github.com/open-telemetry/semantic-conventions/issues/855